src/requestSchemas.ts

```diff
@@ -0,0 +1,169 @@
+  'derived-from',
+  'successor',
+  'predecessor'
+];
+
+const customErrorMap: z.ZodErrorMap = (issue, ctx) => {
+  if (issue.code === z.ZodIssueCode.custom) {
+    if (issue.params?.type === 'NotImplemented') {
+      throw new NotImplementedError(issue.message ?? ctx.defaultError);
+    } else {
+      throw new BadRequestError(issue.message ?? ctx.defaultError, issue.params?.serverErrorCode);
+    }
+  } else if (issue.code === z.ZodIssueCode.unrecognized_keys) {
+    throw new BadRequestError(issue.message ?? ctx.defaultError, constants.ISSUE.CODE.VALUE);
+  }
+  throw new BadRequestError(issue.message ?? ctx.defaultError);
+};
+
+z.setErrorMap(customErrorMap);
+
+const hasIdentifyingInfo = (args: Record<string, any>) => args.id || args.url || args.identifier;
+
+/**
+ * Returns a function that checks if any unsupported params are present, then runs the
+ * other passed in functions in sequence. Each catchFunction is expected to check for
+ * invalid input and call ctx.addIssue if invalid input is detected. Any added issue will
+ * trigger an error to be thrown.
+ */
+export function catchInvalidParams(
+  catchFunctions: ((val: Record<string, any>, ctx: z.RefinementCtx) => void)[],
+  unsupportedArgs?: string[]
+) {
+  return (val: Record<string, any>, ctx: z.RefinementCtx) => {
+    const includedUnsupported = Object.keys(val).filter((v: string) => unsupportedArgs?.includes(v));
+    if (includedUnsupported.length > 0) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        params: { type: 'NotImplemented' },
+        message: `Parameters: ${includedUnsupported.join(', ')} are not yet supported on this server`
+      });
+    } else {
+      catchFunctions.forEach(func => func(val, ctx));
+    }
+  };
+}
+
+/**
+ * Checks that either id, url, or identifier parameter is included. Adds an issue to the ctx
+ * that triggers a BadRequest to be thrown if not.
+ */
+export function catchMissingIdentifyingInfo(val: Record<string, any>, ctx: z.RefinementCtx) {
+  if (!hasIdentifyingInfo(val)) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.custom,
+      params: { serverErrorCode: constants.ISSUE.CODE.REQUIRED },
+      message: 'Must provide identifying information via either id, url, or identifier parameters'
+    });
+  }
+}
+
+const stringToBool = z.enum(['true', 'false']).transform(x => x === 'true');
+const stringToNumber = z.coerce.number();
+const checkDate = (paramName: string) => {
+  return z.string().regex(DATE_REGEX, `${paramName} parameter is not a valid FHIR date`);
+};
+
+export const IdentifyingParameters = z
+  .object({
+    id: z.string(),
+    url: z.string().url(),
+    version: z.string(),
+    identifier: z.string()
+  })
+  .partial();
+
+export const PackageArgs = IdentifyingParameters.extend({
+  capability: z.string(),
+  offset: stringToNumber,
+  count: stringToNumber,
+  'system-version': z.string().url(),
+  'check-system-version': stringToBool,
+  'force-system-version': stringToBool,
```

Comment:
Yes these should not be booleans, but `url` is also technically incorrect. These are type `canonical` in FHIR which supports things like `http://loinc.org|2.56`, which the `|` is not recognized as a valid URL.

if you want to create/use a regex for canonical, go for it, but otherwise anythiing that's canonical should just be `z.string()`